[GeoMechanicsApplication] Avoid slicing of class `StressStatePolicy`

avdg81 opened the pull request:
**📝 Description**
Be more explicit about the intended design of stress state policies. Each policy instance can be moved, but it cannot be copied. To make a deep copy of a stress state policy instance, use member function `Clone`. Unit tests now cover these type traits.

avdg81 committed:
Avoid slicing of class `StressStatePolicy`

avdg81 committed:
Change some `override`s to `final`

The concrete stress state policies should not be used as base classes.

Also, the code base has been reformatted using Clang-format.

avdg81 committed:
Check that all stress state policies are movable, but not copyable

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/axisymmetric_stress_state.h:
```diff
@@ -23,20 +23,20 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) AxisymmetricStressState : public Str
     [[nodiscard]] double CalculateIntegrationCoefficient(const Geometry<Node>::IntegrationPointType& rIntegrationPoint,
                                                          double DetJ,
-                                                         const Geometry<Node>& rGeometry) const override;
+                                                         const Geometry<Node>& rGeometry) const final;
```
I think to ensure that we don't create derived classes we can make the inheritance itself final as well, i.e. `class KRATOS_API(GEO_MECHANICS_APPLICATION) AxisymmetricStressState final : public StressStatePolicy`

see https://en.cppreference.com/w/cpp/language/final

Since in that case we prohibit creating a derived class altogether, I think we could keep the rest `override`. Curious what you think about this!

avdg81 replied:
Agreed, by making the derived classes `final`, we articulate the intend much clearer. Thanks for the suggestion.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/stress_state_policy.h:
```diff
@@ -25,8 +25,17 @@ class Serializer;
+    StressStatePolicy()          = default;
     virtual ~StressStatePolicy() = default;
 
+    // To avoid the slicing problem, copying this class has been prohibited.
+    // Use member `Clone` to make deep copies.
```
Good idea to add this comment explicitly 👍

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_axisymmetric_stress_state.cpp:
```diff
@@ -77,13 +78,21 @@ KRATOS_TEST_CASE_IN_SUITE(ReturnCorrectIntegrationCoefficient, KratosGeoMechanic
+    KRATOS_EXPECT_FALSE(std::is_copy_constructible_v<AxisymmetricStressState>)
+        KRATOS_EXPECT_FALSE(std::is_copy_assignable_v<AxisymmetricStressState>)
+            KRATOS_EXPECT_TRUE(std::is_move_constructible_v<AxisymmetricStressState>)
+                KRATOS_EXPECT_TRUE(std::is_move_assignable_v<AxisymmetricStressState>)}
```
It seems (at least on github), clang-format can't really handle multiple lines of macros. I'd either use the `// clang-format off/on` specifiers, or use EXPECT_TRUE directly, which needs a `;` (therefore I expect clang-format can handle that better). Same holds for the other tests

avdg81 replied:
I've changed it to using `EXPECT_TRUE` and `EXPECT_FALSE` directly, which requires a trailing `;`

rfaasse reviewed (commented):
Thanks for making this more explicit! If there is some time during the demo, it might be nice to use a/some slides to explain why this is necessary, because as discussed, I think all the subtleties within (copy/move)construction semantics, especially in combination with inheritance and user defined versions, are among the trickiest parts of C++ as a language.

I don't have any blockers, just a few formatting suggestions and some thoughts about the final specifier

avdg81 committed:
Processed Richard's review comments

- Marked classes that derive from `StressStatePolicy` `final`.
- Reverted marking the member functions `final`; changed it back to `override`.
- Use GTest's macros `EXPECT_TRUE` and `EXPECT_FALSE` directly, to enforce a better formatting (due to the trailing `;`).

markelov208 reviewed (approved):
Hi Anne, thank you very much for this PR. It is clean and very understandable. I do not have any comments on the implementation.

rfaasse reviewed (approved):
Thanks for processing the review comments, ready to go!
